src/Runner.Worker/CompositeStepsRunner.cs

```diff
@@ -0,0 +1,249 @@
+using GitHub.DistributedTask.Pipelines.ContextData;
+using GitHub.DistributedTask.Pipelines.ObjectTemplating;
+using GitHub.DistributedTask.WebApi;
+using GitHub.Runner.Common;
+using GitHub.Runner.Common.Util;
+using GitHub.Runner.Sdk;
+using GitHub.Runner.Worker.Expressions;
+using ObjectTemplating = GitHub.DistributedTask.ObjectTemplating;
+using Pipelines = GitHub.DistributedTask.Pipelines;
+
+namespace GitHub.Runner.Worker
+{
+    [ServiceLocator(Default = typeof(CompositeStepsRunner))]
+    public interface ICompositeStepsRunner : IRunnerService
+    {
+        Task RunAsync(IExecutionContext Context);
+    }
+
+
+    public sealed class CompositeStepsRunner : RunnerService, ICompositeStepsRunner
+    {
+        public async Task RunAsync(IExecutionContext actionContext)
+        {
+            // Another approach we can explore, is moving all this logic to the CompositeActionHandler if it's small enough. 
+
+            ArgUtil.NotNull(actionContext, nameof(actionContext));
+            ArgUtil.NotNull(actionContext.CompositeSteps, nameof(actionContext.CompositeSteps));
+
+            // The parent StepsRunner of the whole Composite Action Step handles the cancellation stuff already. 
+            while (actionContext.CompositeSteps.Count > 0)
+            {
+                // This is used for testing UI appearance.
+                // System.Threading.Thread.Sleep(5000);
+
+                var step = actionContext.CompositeSteps[0];
+                actionContext.CompositeSteps.RemoveAt(0);
+
+                Trace.Info($"Processing composite step: DisplayName='{step.DisplayName}'");
+
+                step.ExecutionContext.ExpressionValues["steps"] = step.ExecutionContext.StepsContext.GetScope(step.ExecutionContext.ScopeName);
+
+                // Populate env context for each step
+                Trace.Info("Initialize Env context for step");
+#if OS_WINDOWS
+                var envContext = new DictionaryContextData();
+#else
+                var envContext = new CaseSensitiveDictionaryContextData();
+#endif
+
+                // Global env
+                foreach (var pair in step.ExecutionContext.EnvironmentVariables)
+                {
+                    envContext[pair.Key] = new StringContextData(pair.Value ?? string.Empty);
+                }
+
+                // Stomps over with outside step env
+                if (step.ExecutionContext.ExpressionValues.TryGetValue("env", out var envContextData))
+                {
+#if OS_WINDOWS
+                    var dict = envContextData as DictionaryContextData;
+#else
+                    var dict = envContextData as CaseSensitiveDictionaryContextData;
+#endif
+                    foreach (var pair in dict)
+                    {
+                        envContext[pair.Key] = pair.Value;
+                    }
+                }
+
+                step.ExecutionContext.ExpressionValues["env"] = envContext;
+
+
+
+                if (step is IActionRunner actionStep)
```

Comment:
you don't need the check `if (step is IActionRunner actionStep)` since it must be an ActionRunner for composite action.